validator-network/src/network_impl.rs

```diff
@@ -90,102 +113,105 @@ where
     // Looks up the peer ID for a validator public key in the DHT and updates the internal cache
     async fn update_peer_id_cache(&self, validator_id: u16, public_key: &LazyPublicKey) {
         let cache_value = match Self::resolve_peer_id(&self.network, public_key).await {
-            Ok(result) => match result {
-                Some(peer_id) => {
-                    log::trace!(
-                        %peer_id,
-                        validator_id,
-                        %public_key,
-                        "Resolved validator peer ID"
-                    );
-                    CacheState::Resolved(peer_id)
-                }
-                None => {
-                    log::trace!(validator_id, %public_key, "Unable to resolve validator peer ID: Entry not found in DHT");
-                    CacheState::Error(NetworkError::Unreachable)
-                }
-            },
-            Err(e) => {
+            Ok(Some(peer_id)) => {
                 log::trace!(
+                    %peer_id,
                     validator_id,
-                    error = ?e,
+                    %public_key,
+                    "Resolved validator peer ID"
+                );
+                Ok(peer_id)
+            }
+            Ok(None) => {
+                log::trace!(validator_id, %public_key, "Unable to resolve validator peer ID: Entry not found in DHT");
+                Err(())
+            }
+            Err(error) => {
+                log::trace!(
+                    validator_id,
+                    ?error,
                     %public_key,
                     "Unable to resolve validator peer ID: Network error"
                 );
-                CacheState::Error(e)
+                Err(())
             }
         };
-        self.validator_peer_id_cache
+
+        match self
+            .validator_peer_id_cache
             .write()
-            .insert(public_key.compressed().clone(), cache_value);
+            .get_mut(public_key.compressed())
+        {
+            Some(cache_entry) => {
+                if let CacheState::InProgress(prev_peer_id) = *cache_entry {
+                    *cache_entry = match cache_value {
+                        Ok(peer_id) => CacheState::Resolved(peer_id),
+                        Err(()) => CacheState::Error(prev_peer_id),
+                    };
+                } else {
+                    unreachable!("cache state must be \"in progress\"");
```

Comment:
I'd add a comment that this function assumes that the cache entry is updated with an `InProgress` entry before.

Reply:
Done.